### PR TITLE
tiproxy: add traffic replay doc

### PR DESCRIPTION
**Pull request by Oreoxmt**
### What is changed, added or deleted? (Required)

TiProxy v1.3.0 will support traffic replay as an experimental feature.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [x] v8.4 (TiDB 8.4 versions)
- [ ] v8.3 (TiDB 8.3 versions)
- [ ] v8.2 (TiDB 8.2 versions)
- [ ] v8.1 (TiDB 8.1 versions)
- [ ] v7.5 (TiDB 7.5 versions)
- [ ] v7.1 (TiDB 7.1 versions)
- [ ] v6.5 (TiDB 6.5 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions))

### What is the related PR or file link(s)?

- This PR is translated from: https://github.com/pingcap/docs-cn/pull/18522
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: tiproxy/tiproxy-traffic-replay.md [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    The `other_errors` table stores unexpected errors, such as network errors or database connection errors, with the following fields:
+
+    - `err_type`: the type of error, represented by a brief error message. For example, `i/o timeout`.
+    - `sample_err_msg`: the complete error message when the error first occurred.
+    - `sample_replay_time`: the time when the error occurred during replay. You can use this to view error information in the TiDB log file.
+    - `count`: the number of occurrences for this error.
+
+    > **Note:**
+    >
+    > The table structure of `tiproxy_traffic_report` might change in future versions. It is not recommended to directly read data from `tiproxy_traffic_report` in your application or tool development.
```
Comment (qiancai) [inline tiproxy/tiproxy-traffic-replay.md:101]:
```suggestion
    > The table schema of `tiproxy_traffic_report` might change in future versions. It is not recommended to directly read data from `tiproxy_traffic_report` in your application or tool development.
```

File: tiproxy/tiproxy-traffic-replay.md [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+## Test throughput
+
+To test cluster throughput, use the `--speed` option to adjust the replay rate.
+
+For example, `--speed=2` executes SQL statements at twice the rate, halving the total replay time:
```
Comment (qiancai) [inline tiproxy/tiproxy-traffic-replay.md:107]:
```suggestion
For example, `--speed=2` executes SQL statements at twice the rate, reducing the total replay time by half:
```

File: tiproxy/tiproxy-traffic-replay.md [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+```shell
+tiproxyctl traffic replay --host 10.0.1.10 --port 3080 --username="u1" --password="123456" --input="/tmp/traffic" --speed=2
+```
+
+Increasing the replay rate only reduces idle time between SQL statements and does not increase the number of connections. When session idle time is already short, increasing the speed might not effectively improve throughput. In such cases, you can deploy multiple TiProxy instances to replay the same traffic files simultaneously, increasing concurrency to improve throughput.
+
+## View and manage tasks
+
+During capture and replay, tasks automatically stop if unknown errors occur. To view current task progress or error information from the last task, use the [`tiproxyctl traffic show`](/tiproxy/tiproxy-command-line-flags.md#traffic-show) command:
```
Comment (qiancai) [inline tiproxy/tiproxy-traffic-replay.md:117]:
```suggestion
During capture and replay, tasks automatically stop if unknown errors occur. To view the current task progress or error information from the last task, use the [`tiproxyctl traffic show`](/tiproxy/tiproxy-command-line-flags.md#traffic-show) command:
```

File: tiproxy/tiproxy-traffic-replay.md [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+```shell
+tiproxyctl traffic cancel --host 10.0.1.10 --port 3080
+```
+
+For more information, see [`tiproxyctl traffic cancel`](/tiproxy/tiproxy-command-line-flags.md#traffic-cancel).
+
+## Limitations
+
+- TiProxy only supports replaying traffic files captured by TiProxy and does not support other file formats. Therefore, the production cluster must first use TiProxy to capture traffic.
```
Comment (qiancai) [inline tiproxy/tiproxy-traffic-replay.md:150]:
```suggestion
- TiProxy only supports replaying traffic files captured by TiProxy and does not support other file formats. Therefore, make sure to capture traffic from the production cluster using TiProxy first.
```

File: tiproxy/tiproxy-command-line-flags.md [hunk 5]
```diff
@@ -27,12 +27,42 @@ This section lists the flags of the server program `tiproxy`.
...
+# download https://tiup-mirrors.pingcap.com/tiproxy-v1.3.0-linux-amd64.tar.gz 22.51 MiB / 22.51 MiB 100.00% 13.99 MiB/s
+ls `tiup --binary tiproxy`ctl
+# /root/.tiup/components/tiproxy/v1.3.0/tiproxyctl
+```
+
+#### Compile from source code
+
+Compilation environment requirement: [Go](https://golang.org/) 1.21 or later
+
+Compilation procedures: Go to the root directory of the [TiProxy project](https://github.com/pingcap/tiproxy), use the `make` command to compile, and generate `tiproxyctl`.
```
Comment (qiancai) [inline tiproxy/tiproxy-command-line-flags.md:57]:
```suggestion
Compilation procedures: Go to the root directory of the [TiProxy project](https://github.com/pingcap/tiproxy), use the `make` command to compile and generate `tiproxyctl`.
```

File: tiproxy/tiproxy-command-line-flags.md [hunk 6]
```diff
@@ -121,3 +156,87 @@ Example output:
...
+#### `traffic replay`
+
+The `tiproxyctl traffic replay` command is used to replay captured traffic.
+
+Options:
+
+- `--username`: (required) specifies the database username for replay.
+- `--password`: (optional) specifies the password for the username. The default value is an empty string `""`.
+- `--input`: (required) specifies the directory containing traffic files.
+- `--speed`: (optional) specifies the replay speed multiplier. The range is `[0.1, 10]`. The default value of 1 indicates replay at original speed.
```
Comment (qiancai) [inline tiproxy/tiproxy-command-line-flags.md:186]:
```suggestion
- `--speed`: (optional) specifies the replay speed multiplier. The range is `[0.1, 10]`. The default value is `1`, indicating replay at the original speed.
```

File: tiproxy/tiproxy-command-line-flags.md [hunk 7]
```diff
@@ -121,3 +156,87 @@ Example output:
...
+
+Example:
+
+The following command connects to the TiProxy instance at `10.0.1.10:3080` using username `u1` and password `123456`, reads traffic files from the `/tmp/traffic` directory on the TiProxy instance, and replays traffic at twice the original speed:
```
Comment (qiancai) [inline tiproxy/tiproxy-command-line-flags.md:190]:
```suggestion
The following command connects to the TiProxy instance at `10.0.1.10:3080` using username `u1` and password `123456`, reads traffic files from the `/tmp/traffic` directory on the TiProxy instance, and replays the traffic at twice the original speed:
```

File: tiproxy/tiproxy-command-line-flags.md [hunk 8]
```diff
@@ -121,3 +156,87 @@ Example output:
...
+
+#### `traffic show`
+
+The `tiproxyctl traffic show` command is used to display historical capture and replay tasks.
+
+The `status` field in the output indicates the task status, with the following possible values:
+
+- `done`: task completed normally.
+- `canceled`: task was canceled. You can check the `error` field for the reason.
+- `running`: task is running. You can check the `progress` field for  completion percentage.
```
Comment (qiancai) [inline tiproxy/tiproxy-command-line-flags.md:208]:
```suggestion
- `done`: the task completed normally.
- `canceled`: the task was canceled. You can check the `error` field for the reason.
- `running`: the task is running. You can check the `progress` field for the completion percentage.
```

**Review by qiancai — APPROVED** [review]
Rest LGTM

File: tiproxy/tiproxy-traffic-replay.md [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+    For more information, see [`tiproxyctl traffic replay`](/tiproxy/tiproxy-command-line-flags.md#traffic-replay).
+
+5. View the replay report.
+
+    After replay completion, the report is stored in the `tiproxy_traffic_report` database on the test cluster. This database contains two tables: `fail` and `other_errors`.
+
+    The `fail` table stores failed SQL statements, with the following fields:
+
+    - `cmd_type`: the type of a failed SQL statement, such as `Query` (execute an ordinary statement), `Prepare` (prepare a statement), and `Execute` (execute a prepared statement).
```
Comment (djshow832) [inline tiproxy/tiproxy-traffic-replay.md:83]:
```suggestion
    - `cmd_type`: the type of a failed command, such as `Query` (execute an ordinary statement), `Prepare` (prepare a statement), and `Execute` (execute a prepared statement).
```

File: tiproxy/tiproxy-traffic-replay.md [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+
+<img src="https://download.pingcap.com/images/docs/tiproxy/tiproxy-traffic-replay.png" alt="TiProxy traffic replay" width="800" />
+
+## Use cases
+
+Traffic replay is suitable for the following scenarios:
+
+- **Validate TiDB version upgrades**: Replay production traffic on a test cluster with a new TiDB version to verify if the new TiDB version can successfully execute all SQL statements.
+- **Assess change impact**: Simulate production traffic on a test cluster to validate the impact of changes on the cluster. For example, verify the effects before modifying configuration items or system variables, altering table schemas, or enabling new TiDB features.
+- **Validate performance before TiDB scaling**: Replay traffic at corresponding rates on a test cluster with a new scale to verify if the performance meets requirements. For example, to plan a 50% cluster downscale for cost savings, replay traffic at half speed to verify if SQL latency meets requirements after scaling.
```
Comment (djshow832) [inline tiproxy/tiproxy-traffic-replay.md:22]: I assume that `Validate` means `check the validation of`, does it also have the meaning of `verify`?

File: tiproxy/tiproxy-traffic-replay.md [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+```shell
+tiproxyctl traffic cancel --host 10.0.1.10 --port 3080
+```
+
+For more information, see [`tiproxyctl traffic cancel`](/tiproxy/tiproxy-command-line-flags.md#traffic-cancel).
+
+## Limitations
+
+- TiProxy only supports replaying traffic files captured by TiProxy and does not support other file formats. Therefore, make sure to capture traffic from the production cluster using TiProxy first.
+- TiProxy traffic replay does not support filtering SQL types and DML and DDL statements are replayed. Therefore, you need to restore the cluster data to its pre-replay state before replaying.
```
Comment (djshow832) [inline tiproxy/tiproxy-traffic-replay.md:151]:
```suggestion
- TiProxy traffic replay does not support filtering SQL types and DML and DDL statements are replayed. Therefore, you need to restore the cluster data to its pre-replay state before replaying again.
```

**Review by djshow832 — APPROVED** [review]

**Pull request merged**